### PR TITLE
Type annotation bundle (logger, vec env, custom envs)

### LINKED ISSUE #1318: Use `Sequence` type hint for `VecEnv.seed` return
Just for me to remember:

https://github.com/DLR-RM/stable-baselines3/blob/411ff697dde31df9ff914ba0c335875d1bd5998d/stable_baselines3/common/vec_env/base_vec_env.py#L196

Should be typed hinted with `Sequence` instead of `List` as `Sequence` is invariant.

### PR DESCRIPTION
**Pull request by araffin**
## Motivation and Context

- [ ] I have raised an issue to propose this change ([required](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) for new features and bug fixes)

closes #1318 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)

## Checklist

- [x] I've read the [CONTRIBUTION](https://github.com/DLR-RM/stable-baselines3/blob/master/CONTRIBUTING.md) guide (**required**)
- [ ] I have updated the changelog accordingly (**required**).
- [ ] My change requires a change to the documentation.
- [ ] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.
- [ ] I have opened an associated PR on the [SB3-Contrib repository](https://github.com/Stable-Baselines-Team/stable-baselines3-contrib) (if necessary)
- [ ] I have opened an associated PR on the [RL-Zoo3 repository](https://github.com/DLR-RM/rl-baselines3-zoo) (if necessary)
- [x] I have reformatted the code using `make format` (**required**)
- [x] I have checked the codestyle using `make check-codestyle` and `make lint` (**required**)
- [x] I have ensured `make pytest` and `make type` both pass. (**required**)
- [x] I have checked that the documentation builds using `make doc` (**required**)

Note: You can run most of the checks using `make commit-checks`.

Note: we are using a maximum length of 127 characters per line

### CODE REVIEW

**Comment by qgallouedec** [comment]
‎️‍🔥 tell me when you're done, I'll be happy to review!

**Comment by araffin** [comment]
Merci =)
i should be done for this bundle.

**Comment by araffin** [comment]
i'll fix the error tmr

File: stable_baselines3/common/vec_env/vec_extract_dict_obs.py [hunk 1]
```diff
@@ -13,14 +14,19 @@ class VecExtractDictObs(VecEnvWrapper):
...
     def __init__(self, venv: VecEnv, key: str):
         self.key = key
+        assert isinstance(
+            venv.observation_space, spaces.Dict
+        ), f"VecExtractDictObs can only be used with Dict obs space, not {venv.observation_space}"
         super().__init__(venv=venv, observation_space=venv.observation_space.spaces[self.key])
 
     def reset(self) -> np.ndarray:
         obs = self.venv.reset()
+        assert isinstance(obs, dict)
```
Comment (qgallouedec) [inline stable_baselines3/common/vec_env/vec_extract_dict_obs.py:24]: in the future, we should probably use [`Generic`](https://docs.python.org/3/library/typing.html#typing.Generic) for `VecEnvWrapper` and `VecEnvWrapper` to handle this case.

File: stable_baselines3/common/vec_env/vec_frame_stack.py [hunk 2]
```diff
@@ -31,13 +31,13 @@ def step_wait(
         self,
     ) -> Tuple[Union[np.ndarray, Dict[str, np.ndarray]], np.ndarray, np.ndarray, List[Dict[str, Any]],]:
         observations, rewards, dones, infos = self.venv.step_wait()
-        observations, infos = self.stacked_obs.update(observations, dones, infos)
+        observations, infos = self.stacked_obs.update(observations, dones, infos)  # type: ignore[arg-type]
```
Comment (qgallouedec) [inline stable_baselines3/common/vec_env/vec_frame_stack.py:34]: Same remark here

**Review by qgallouedec — APPROVED** [review]

File: stable_baselines3/common/vec_env/vec_extract_dict_obs.py (hunk 1 above)
Comment (araffin) [reply stable_baselines3/common/vec_env/vec_extract_dict_obs.py:24]: I agree, my goal for now is to remove as many files as possible from mypy ignore and then do a second pass later to polish the edges (i.e. better typing hints).

File: stable_baselines3/common/vec_env/vec_monitor.py [hunk 3]
```diff
@@ -23,9 +23,6 @@ class VecMonitor(VecEnvWrapper):
     :param info_keywords: extra information to log, from the information return of env.step()
     """
 
-    episode_returns: np.ndarray
```
Comment (araffin) [inline stable_baselines3/common/vec_env/vec_monitor.py]: I usually do that to declare explicitly members that are not defined in the `__init__`.
The goal is that the user knows easily what are the members of a given class by looking only at the `__init__`.
Or is there a better way?

Comment (qgallouedec) [reply stable_baselines3/common/vec_env/vec_monitor.py]: I see, I thought it was only to satisfy mypy.
I don't know a better way. I think you have to look at it on a case-by-case basis. Here for example, we could simply add it in the init:

```python
class VecMonitor(VecEnvWrapper):
     def __init__(self, ...):
        ...
        self.episode_returns = np.zeros(self.num_envs, dtype=np.float32)
        self.episode_lengths = np.zeros(self.num_envs, dtype=np.int32)
```

or keep what you had originally done. As you like!

**Pull request merged**